src/components/utils.ts

```diff
@@ -0,0 +1,65 @@
+import { FlatOption, Option } from "./types";
+
+export const flattenOptions = (array: FlatOption[], options: Option[], path: string[] = []) => {
+    options.forEach(op => {
+        const hasChildren = op.hasOwnProperty("children");
+        const newPath = [...path, op.id];
+        const isTopLevel = newPath.length === 1;
+
+        const baseOption = {
+            id: op.id,
+            label: op.label,
+            path: newPath,
+            show: isTopLevel
+        };
+
+        if (hasChildren) {
+            array.push({...baseOption, hasChildren, open: false});
+            flattenOptions(array, op.children!, newPath);
+        } else {
+            array.push({...baseOption, hasChildren});
+        }
+    });
+};
+
+const arraysAreEqual = (array1: string[], array2: string[]) => {
+    return JSON.stringify(array1) === JSON.stringify(array2);
+};
+
+export const expandOptions = (array: FlatOption[], optionPath: string[]): FlatOption[] => {
```

Comment:
I do wonder if it would be a bit more Vue-y to modify the array in place rather than map to a whole new array. I "think" that should work if the FlatOption array was created as `reactive` when initialised.

Reply:
Or maybe `ref` would be fine - can never remember how it treats arrays, but I'm sure it can handle changing arrays!

Reply:
ref was indeed fine XD